c10/xpu/impl/XPUGuardImpl.h

```diff
@@ -0,0 +1,130 @@
+#pragma once
+
+#include <c10/core/DeviceGuard.h>
+#include <c10/core/impl/DeviceGuardImplInterface.h>
+#include <c10/xpu/XPUCachingAllocator.h>
+#include <c10/xpu/XPUFunctions.h>
+#include <c10/xpu/XPUStream.h>
+
+#include <vector>
+
+namespace c10::xpu::impl {
+
+struct XPUGuardImpl final : public c10::impl::DeviceGuardImplInterface {
+  static constexpr DeviceType static_type = DeviceType::XPU;
+
+  XPUGuardImpl() = default;
+
+  explicit XPUGuardImpl(DeviceType t) {
+    TORCH_INTERNAL_ASSERT(t == DeviceType::XPU);
+  }
+
+  DeviceType type() const override {
+    return DeviceType::XPU;
+  }
+
+  Device exchangeDevice(Device d) const override {
+    TORCH_INTERNAL_ASSERT(d.is_xpu());
+    auto old_device_index = c10::xpu::exchange_device(d.index());
+    return Device(DeviceType::XPU, old_device_index);
+  }
+
+  Device getDevice() const override {
+    auto device = c10::xpu::current_device();
+    return Device(DeviceType::XPU, device);
+  }
+
+  c10::optional<Device> uncheckedGetDevice() const noexcept {
```

Comment:
Do we need this since it's not on the interface?

Reply:
This is aligned with CUDA. We intend to have our APIs align one-to-one with CUDA, saving the effort for users to migrate their code.

Reply:
There is litterally 0 use of this function across all of GitHub: https://github.com/search?q=%2FuncheckedGetDevice%2F&type=code

Reply:
Thanks, removed.